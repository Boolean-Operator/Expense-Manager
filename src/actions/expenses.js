import uuid from 'uuid';
import database from '../firebase/firebase';

// Expense Action Generators
// Components call action generators
// Action generators return function
// Components dispatch function
// Function runs ( dispatch other actions )

//  ADD_EXPENSE
const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense: expense
});

// starts Add Expense Action, 
const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {description, note, amount, createdAt}

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    });
  };
};

//  REMOVE_EXPENSE
const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id: id
});

//starts Remove Expense Action
const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`)
        .remove()
        .then(() => {
          dispatch(removeExpense({ id }))
        }).catch((error) => {
          console.log("Dave, I afraid I can't do that.", error);
        });
  }
}

//  EDIT_EXPENSE
const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id: id,
  updates: updates
});

//  SET_EXPENSES
const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// fetch all expense data once
// Parse that data to array
// dispatch expenses, instead of c.log

const startSetExpenses = () => {
  return (dispatch) => {

  // Fetch all expense data once
  // Parse data into array
  // Dispatch SET_EXPENSES
    return database.ref('expenses')
      .once('value')
      .then((snapshot) => {
        const expenses = [];
    
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
    
      }).catch((error) => {
        console.log('Dave, I afraid something has gone terribly wrong. ', error);
      });
  }
};




export {
  addExpense,
  startAddExpense,
  removeExpense,
  startRemoveExpense,
  editExpense,
  setExpenses,
  startSetExpenses
}