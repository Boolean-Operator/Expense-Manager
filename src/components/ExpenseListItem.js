import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';

 //Expense List Item component

const ExpenseListItem = ({id, description, amount, createdAt }) => (
  <div>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    <p>${amount} -- Incurred:{moment(createdAt).format("MMM Do YYYY")}</p>
  </div>
);

export default ExpenseListItem;
