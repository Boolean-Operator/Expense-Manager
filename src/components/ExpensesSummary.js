// Expenses Summary Component

// Render in ExpenseDashboardPage

//Test with two snapshots
// Viewing 1 expense totalling $94.34 
// Viewing 2 expenses totalling $94.34

// Connect to store for:
  // expensesCount (of visible expenses)
  // expensesTotal (of visible expenses)

// Commit and Deploy to Heroku


import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedExpensesTotal  = numeral(expensesTotal / 100).format('$0,0.00');
 
  return (
    <div>
      <h2>
        Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}
      </h2>
    </div>
  );
};
 
const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  }

}

export default connect(mapStateToProps)(ExpensesSummary)

