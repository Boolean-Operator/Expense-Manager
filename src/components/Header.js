import React from 'react';
import { NavLink } from 'react-router-dom';


const Header = () => (
  <header>
    <h1>Expensify Header</h1>
    <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard-*-</NavLink>
    <NavLink to="/create" activeClassName="is-active">Add Expense-*-</NavLink>
  </header>
)
export default Header;