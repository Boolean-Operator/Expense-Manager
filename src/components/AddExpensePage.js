import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

//Add Expense Component
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/')
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
        </h1>
      </div>    
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
   startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

  export default connect(undefined, mapDispatchToProps)(AddExpensePage);