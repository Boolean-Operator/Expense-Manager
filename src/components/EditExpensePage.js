import React from 'react';
import { connect } from 'react-redux';
import  ExpenseForm from './ExpenseForm'
import { editExpense, removeExpense } from '../actions/expenses'

// Edit Expense Component
export class EditExpensePage extends React.Component {
  
  onSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onClick = () => {
    this.props.removeExpense({id:this.props.expense.id});
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <h1>Edit Expense
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button
          expense={this.props.expense}
          onClick={this.onClick}>Remove</button>
        </h1>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
});
    
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);