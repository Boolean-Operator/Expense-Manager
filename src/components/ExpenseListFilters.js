import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount , sortByDate, setStartDate, setEndDate } from '../actions/filters'
import { DateRangePicker } from 'react-dates'

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate),
    this.props.setEndDate(endDate)
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value)
  };
  onSortChange = (e) => {
    (e.target.value === 'date') ?
      this.props.sortByDate() :
      this.props.sortByAmount();
  };
  render() {
    return (
      <div>
        <input
          type='text'
          value={this.props.filters.text}
          onChange={this.onTextChange} 
        />
        <select 
          value={this.props.filters.sortBy}
          onChange={this.onSortChange}>
          <option value='date'>Date</option>
          <option value='amount'>Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          // startDateId={}
          endDate={this.props.filters.endDate}
          // endDateId={}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}


        />
      </div>
    );

  }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);