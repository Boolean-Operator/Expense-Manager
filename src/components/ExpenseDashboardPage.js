import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters'
import ExpensesSummary from './ExpensesSummary';

// remove h1 tags
const ExpenseDashboardPage = () => (
  <div>
    <h1>
      <ExpensesSummary />
      <ExpenseListFilters />
      <ExpenseList />
    </h1>
  </div>
);

export default ExpenseDashboardPage; 