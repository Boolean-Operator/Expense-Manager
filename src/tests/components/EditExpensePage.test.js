import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { EditExpensePage } from '../../components/EditExpensePage';

let editExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
  editExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      editExpense={editExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense = {expenses[1]}
    />
  );
});

test('should render Edit Expense Page correctly', ()  => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

test('should handle startRemoveExpense', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(startRemoveExpense).toHaveBeenLastCalledWith({
    id: expenses[1].id
  });
});