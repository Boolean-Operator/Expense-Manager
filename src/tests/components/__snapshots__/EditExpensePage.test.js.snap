// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Edit Expense Page correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "description": "Rent",
        "id": "2",
        "note": "",
      }
    }
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
